### PR TITLE
verify_doc.py: verification of API documentation

### PR DESCRIPTION
**Pull request by lior-stratoscale**
@Stratoscale/mgmt please review

### CODE REVIEW

File: hammock/verify_doc.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    return errors
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('package')
+    args = parser.parse_args()
+    parser.add_mutually_exclusive_group()
+    errors = verify_api(args.package)
+    print('\n'.join(errors))
```
Comment (maroun-stratoscale) [inline hammock/verify_doc.py]: Someone is getting ready for Python3.

**Comment by ravid-stratoscale** [comment]
I don't know hammock enough to verify that, maybe @eyal-stratoscale can

File: hammock/verify_doc.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+def _func_default_dict(route):
+    func_args = inspect.getargspec(route.func).args
+    arg_names = [route.keyword_map.get(name, name) for name in func_args]
+    defaults = inspect.getargspec(route.func).defaults or tuple()
+    defaults = (NO_DEFAULT, ) * (len(arg_names) - len(defaults)) + defaults
+    return dict(zip(arg_names, defaults))
+
+
+def verify_doc(package, method_verifier=None, argument_verifier=None, verification_exceptions=None):
```
Comment (yossi-stratoscale) [inline hammock/verify_doc.py]: Kinda funny that's the doc verifier is not documented

File: hammock/verify_doc.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    verification_exceptions = verification_exceptions or set()
+    errors = []
+    with mock_import.mock_import([package]):
+        for resource_class, parents in packages.iter_resource_classes(package):
+            resource_path = common.url_join(*([parent.path for parent in parents] + [resource_class.path()]))
+            for route in resource_class.iter_route_methods():
+                if route.dest is not None:
+                    continue
+                path = '/' + common.url_join(resource_path, route.path)
+                identifier = '%s::%s' % (resource_class.__name__, route.func.__name__)
```
Comment (eyal-stratoscale) [inline hammock/verify_doc.py]: I think in python `::` should be `.`

File: hammock/verify_doc.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                                success_code=route.success_code)
+                errors.extend(_verify_method(identifier, method, method_verifier))
+
+                default_dict = _func_default_dict(route)
+                for name, arg in six.iteritems(route.spec.args_info):
+                    if name.startswith('_'):
+                        continue
+                    if (resource_class.__name__, route.func.__name__, route.keyword_map.get(name, name)) in verification_exceptions:
+                        continue
+                    identifier = '%s::%s' % (resource_class.__name__, route.func.__name__)
```
Comment (eyal-stratoscale) [inline hammock/verify_doc.py]: appears twice so maybe should be a function

File: hammock/verify_doc.py (hunk 2 above)
Comment (eyal-stratoscale) [inline hammock/verify_doc.py]: agree, it would be nice to have a bit documentation in the code

File: hammock/client.py [hunk 5]
```diff
@@ -23,7 +24,18 @@
 
 
 class ClientGenerator(object):
-    def __init__(self, class_name, resources_package, default_url=''):
+    def __init__(self, class_name, resources_package, default_url='',
+                 lenient=False, method_verifier=False, argument_verifier=False,
+                 verification_exceptions=None):
```
Comment (eyal-stratoscale) [inline hammock/client.py]: Not clear to me what does `verification_exceptions` means, and why is it in the client generator class init function.
Can you please explain/document? a list of tuples of what?
Thanks

File: hammock/verify_doc.py [hunk 6]
```diff
@@ -67,14 +67,34 @@ def _func_default_dict(route):
...
+    object (see above) and returns a list of errors as strings (if
+    found).  these strings are prefixed by the method identifier and
+    added to any other errors that are found here.
+
+    :param argument_verifier: a function that gets a single Argument
+    object (see above) and returns a list of errors as strings (if
+    found).  these strings are prefixed by the argument identifier and
+    added to any other errors that are found here.
+
+    :verification_exceptions: a list of tuples for methods and
```
Comment (eyal-stratoscale) [inline hammock/verify_doc.py]: I see, maybe use a class for that? it is too complicated to my opinion

**Pull request merged**